File: learn-algo/leetcode/748/c++/748.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    string shortestCompletingWord(string licensePlate, vector<string> &words) {
        // 统一格式化为小写
        transform(licensePlate.begin(), licensePlate.end(), licensePlate.begin(), ::tolower);

        unordered_map<char, int> kMap;
        for_each(licensePlate.begin(), licensePlate.end(), [&kMap](char c) { if (isalpha(c)) kMap[c]++; });

//        for_each(kMap.begin(), kMap.end(), [](auto t){cout << t.first << " " << t.second << endl; });

        int minLenIndex = words.size() - 1;
        int minLen = INT_MAX;

        for (auto i = 0; i < words.size(); i++) {
            unordered_map<char, int> tMap;
            bool isSuccess = true;
            auto word = words[i];
            for_each(word.begin(), word.end(), [&tMap](char c) { if (isalpha(c)) tMap[c]++; });
            for (auto v: kMap) {
                if (tMap[v.first] < v.second) {
                    isSuccess = false;
                    break;
                }
            }

            if (!isSuccess) {
                continue;
            }

            if (word.size() < minLen) {
                minLenIndex = i;
                minLen = word.size();
            }
        }

        return words[minLenIndex];

    }
};

int main() {
    Solution s;
    auto licensePlate = "1s3 PSt";
    vector<string> words = {"step", "steps", "stripe", "stepple"};

    auto res = s.shortestCompletingWord(licensePlate, words);
    cout << res << endl;
    return 0;
}

File: learn-algo/leetcode/748/go/748.go
package main

import (
	"fmt"
	"strings"
)

const INTMAX = int(^uint(0) >> 1)

func isAlpha(c byte) bool {
	if isUpAlpha(c) || isLowAlpha(c) {
		return true
	}
	return false
}

func isUpAlpha(c byte) bool {
	if c >= 65 && c <= 90 {
		return true
	}
	return false
}

func isLowAlpha(c byte) bool {
	if c >= 97 && c <= 122 {
		return true
	}
	return false
}

func isDigit(c byte) bool {
	if c >= 48 && c <= 57 {
		return true
	}
	return false
}

func shortestCompletingWord(licensePlate string, words []string) string {
	licensePlate = strings.ToLower(licensePlate)

	var getMap = func(s string) map[uint8]int {
		var kMap = make(map[byte]int)
		for _, v := range []byte(s) {
			if isAlpha(v) {
				kMap[v]++
			}
		}
		return kMap
	}

	kMap := getMap(licensePlate)

	var isTrue = func(uMap map[byte]int) bool {
		for k, v := range kMap {
			if uMap[k] < v {
				return false
			}
		}
		return true
	}

	var minLen = INTMAX
	var minLenIndex = len(licensePlate) - 1

	for i, word := range words {
		tMap := getMap(word)
		if isTrue(tMap) {
			if len(word) < minLen {
				minLen = len(word)
				minLenIndex = i
			}
		}
	}

	return words[minLenIndex]
}

func main() {
	var lic = string("Ah71752")
	var words = []string{"of", "husband", "easy", "education", "drug", "prevent", "writer", "old"}
	res := shortestCompletingWord(lic, words)
	fmt.Println(res)

}
